Segmentation/data_augmentation/GAN_try.py

```diff
@@ -0,0 +1,134 @@
+
+# #showing image in mnist
+# plt.imshow(x_train[10])
+# plt.show()
+
+BATCH_SIZE = 256
+
+def generator_model(noise_len):
+    model = tf.keras.Sequential()
+
+    #Adding first dense layer
+    model.add(layers.Dense(7*7*BATCH_SIZE, use_bias = False, input_shape=(noise_len, )))
+    model.add(layers.BatchNormalization())
+    model.add(layers.LeakyReLU())
+
+    model.add(layers.Reshape((7, 7, BATCH_SIZE)))
+    assert model.output_shape == (None, 7, 7, BATCH_SIZE)
+
+    #First transpose convolutional layer
+    model.add(layers.Conv2DTranspose(128, (5,5), strides=1, padding='same', use_bias=False))
+    model.add(layers.BatchNormalization())
+    model.add(layers.LeakyReLU())
+
+    #Second transpose convolutional layer
+    model.add(layers.Conv2DTranspose(64, (5,5), strides=2, padding='same', use_bias=False))
+    model.add(layers.BatchNormalization())
+    model.add(layers.LeakyReLU())
+
+    #Last transpose convolutional layer
+    model.add(layers.Conv2DTranspose(1, (5,5), strides=2, padding='same', use_bias=False, activation='tanh'))
+    assert model.output_shape == (None, 28, 28, 1)
+
+    return model
+
+#Test the genrator with random noise
+noise_length = 100
+
+generator_tryingOut = generator_model(noise_length)
+generator_tryingOut.summary()
+
+noise = tf.random.normal([1, noise_length])
+generated_image = generator_tryingOut(noise, training=False)
+
+plt.imshow(generated_image[0,:,:,0], cmap='gray')
+plt.show()
+
+def discriminator_model():
+    model = tf.keras.Sequential()
+
+    #Fist convolutional layer
+    model.add(layers.Conv2D(64, 5, strides=2, padding='same', input_shape=[28,28,1]))
+    model.add(layers.LeakyReLU())
+    model.add(layers.Dropout(0.3))
+
+    #Second covolutional layer
+    model.add(layers.Conv2D(128, 5, strides=2, padding='same'))
+    model.add(layers.LeakyReLU())
+    model.add(layers.Dropout(0.3))
+
+    #Dense Layer for output --> therefore flatten
+    model.add(layers.Flatten())
+    model.add(layers.Dense(1))
+
+    return model
+
+discriminator_tryingOut = discriminator_model()
+decision = discriminator_tryingOut(generated_image)
+print(decision)
+
+#Loss functions
+cross_entropy = tf.keras.losses.BinaryCrossentropy(from_logits=True)  #Defining the cross entropy loss function
+
+def discriminator_loss(real_ouput, fake_output):
+    real_loss = cross_entropy(tf.ones_like(real_ouput), real_ouput)
+    fake_loss = cross_entropy(tf.zeros_like(fake_output), fake_output)
+    total_loss = real_loss + fake_loss
+    return total_loss
+
+def generator_loss(fake_output):
+    return cross_entropy(tf.ones_like(fake_output), fake_output)
+
+#Optimizers
+generator_optimizer = tf.keras.optimizers.Adam(1e-4)
+discriminator_optimizer = tf.keras.optimizers.Adam(1e-4)
+
+#Training
+epochs = 2
+number_of_generated_output = 16
+noise_length = 100
+#seed = tf.random.normal([number_of_generated_output, noise_length])
+
+generator = generator_model(noise_length)
+discriminator = discriminator_model()
+
```

Comment:
Don't forget to add metrics for loss and the discriminator's accuracy. Need to have an option to track the losses while the model's training.